pkg/operator/controller/canary/route.go

```diff
@@ -75,6 +77,21 @@ func (r *reconciler) updateCanaryRoute(current, desired *routev1.Route) (bool, e
 		return false, nil
 	}
 
+	if len(updated.Spec.Host) == 0 && len(current.Spec.Host) != 0 {
+		// Attempts to clear spec.host may be ignored.  Thus, to clear
+		// spec.host, it is necessary to delete and recreate the route.
+		log.Info("deleting and recreating the canary route to clear spec.host", "namespace", current.Namespace, "name", current.Name, "old spec.host", current.Spec.Host)
+		foreground := metav1.DeletePropagationForeground
```

Comment:
`DeletePropagationForeground` [keeps the key-value in the store until all the dependencies are also deleted (making sure there are no orphans)](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L489-L494).  It's not perfectly clear to me that this is needed, but I don't expect it to cause any harm.

Reply:
The godoc for the `DeletePropagationForeground` value is getting into the implementation details, but [the godoc for the `PropagationPolicy` field](https://github.com/kubernetes/apimachinery/blob/6b362fab6df25da6dc077a5414bb1e54235a8bc7/pkg/apis/meta/v1/types.go#L530-L538) is more to the point: 
```go
	// Whether and how garbage collection will be performed.
	// Either this field or OrphanDependents may be set, but not both.
	// The default policy is decided by the existing finalizer set in the
	// metadata.finalizers and the resource-specific default policy.
	// Acceptable values are: 'Orphan' - orphan the dependents; 'Background' -
	// allow the garbage collector to delete the dependents in the background;
	// 'Foreground' - a cascading policy that deletes all dependents in the
	// foreground.
```
The important thing is that deletion happens in the foreground so that the operator can immediately recreate the route.